### PR TITLE
Zip v0.4

### PR DESCRIPTION
**Pull request by hzhuang1**

### CODE REVIEW

**Review by wangzhou — COMMENTED** [review]
This is done by Jean, please make sure if he is still using this one ：）

File: test/hisi_zip_test/test_lib.c [hunk 1]
```diff
@@ -569,43 +569,75 @@ static void *poll_thread_func(void *arg)
...
+				i, ret);
+			return ret;
+		}
 	}
-	ret = pthread_create(&info->threads[1], &attr, send_thread_func, info);
-	if (ret < 0)
-		return ret;
 	pthread_attr_destroy(&attr);
 	g_conf = &info->ctx_conf;
 	return 0;
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c:1273]: should add some error handle, like memory free, thread destroy, global variables unset

File: test/hisi_zip_test/test_lib.c [hunk 2]
```diff
@@ -569,43 +569,75 @@ static void *poll_thread_func(void *arg)
...
+				     poll_thread_func, info);
+		if (ret < 0) {
+			fprintf(stderr, "Fail to create send thread %d (%d)\n",
+				i, ret);
+			return ret;
+		}
+	}
+	pthread_attr_destroy(&attr);
+	count = 0;
+	return 0;
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c:1366]: same as here

File: test/hisi_zip_test/test_lib.h [hunk 3]
```diff
@@ -111,7 +113,13 @@ struct hizip_test_info {
...
 
 void *send_thread_func(void *arg);
-int create_threads(struct hizip_test_info *info);
+void *poll_thread_func(void *arg);
+int create_send_threads(struct hizip_test_info *info,
+			void *(*send_thread_func)(void *arg),
+			int num);
+int create_poll_threads(struct hizip_test_info *info,
+			void *(*poll_thread_func)(void *arg),
+			int num);
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.h:174]: And logically, there should have the "free" functions about create_send_threads and create_poll_threads

**Review by wangzhou — COMMENTED** [review]
test: zip: fix to create session in each thread ?

s/earch/each/

**Review by wangzhou — COMMENTED** [review]
test: zip: fix ctx mode and type unmatched with sched
Although sched requires different mode and type, mode and type and
hacked to single value. Fix it.

this "mode and type and hacked to single value."  s/and/are/  ？

File: test/hisi_zip_test/test_lib.c [hunk 4]
```diff
@@ -713,75 +713,89 @@ int init_ctx_config(struct test_options *opts, void *priv,
...
+		WD_ERR("Not enough memory to allocate contexts.\n");
+		ret = -ENOMEM;
+		goto out_ctx;
+	}
+	for (i = 0; i < ctx_conf->ctx_num; i++) {
+		ctx_conf->ctxs[i].ctx = wd_request_ctx(info->list->dev);
+		if (!ctx_conf->ctxs[i].ctx) {
+			WD_ERR("Fail to allocate context #%d\n", i);
+			ret = -EINVAL;
+			goto out_req;
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c:1537]: This error handle is wrong, as we do not free allocated ctx here.

**Review by wangzhou — COMMENTED** [review]
test: zip: fix ctx mode and type unmatched with sched
Although sched requires different mode and type, mode and type and
hacked to single value. Fix it.

this "mode and type and hacked to single value."  s/and/are/  ？

File: test/hisi_zip_test/test_sva_perf.c [hunk 5]
```diff
@@ -395,8 +395,9 @@ static int run_one_test(struct test_options *opts, struct hizip_stats *stats)
 
 	stat_start(&info);
 	create_send_threads(&info, send_thread_func, opts->thread_num);
-	create_poll_threads(&info, poll_thread_func, 1);
-	attach_threads(&info);
+	if (opts->sync_mode)
```
Comment (wangzhou) [inline test/hisi_zip_test/test_sva_perf.c:402]: hide this in create_poll_threads ?

File: test/hisi_zip_test/test_sva_perf.c [hunk 6]
```diff
@@ -419,8 +420,9 @@ static int run_one_test(struct test_options *opts, struct hizip_stats *stats)
 			fprintf(stderr, "NOTE: test might trash the TLB\n");
 
 		create_send_threads(&info, send_thread_func, opts->thread_num);
-		create_poll_threads(&info, poll_thread_func, 1);
-		ret = attach_threads(&info);
+		if (opts->sync_mode)
+			create_poll_threads(&info, poll_thread_func, 1);
```
Comment (wangzhou) [inline test/hisi_zip_test/test_sva_perf.c:428]: same as above

File: test/hisi_zip_test/test_lib.c [hunk 7]
```diff
@@ -581,27 +582,40 @@ int create_send_threads(struct test_options *opts,
 			)
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c:1243]: this is no the problem of this patch. However, ) should be in above line :)

File: test/hisi_zip_test/test_sva_perf.c [hunk 8]
```diff
@@ -732,6 +733,54 @@ static void handle_sigbus(int sig)
...
+	       "at %.2fMB/s in %f usec.\n", opts.thread_num, speed, usec);
+	free_threads(&info);
+	return 0;
+}
+
+static int run_self_test(void)
+{
+	int ret, f_ret = 0;
+
+	printf("Start to run self test!\n");
```
Comment (wangzhou) [inline test/hisi_zip_test/test_sva_perf.c]: suggest to update all print with a common prefix, like uadk comp test: xxx

File: test/hisi_zip_test/test_sva_perf.c [hunk 9]
```diff
@@ -732,6 +733,54 @@ static void handle_sigbus(int sig)
...
+	speed = ilen * 1000 * 1000 / 1024 / 1024 / usec;
+	printf("Mixture of SW compress and SW decompress with %d threads "
+	       "at %.2fMB/s in %f usec.\n", opts.thread_num, speed, usec);
+	free_threads(&info);
+	return 0;
+}
+
+static int run_self_test(void)
+{
+	int ret, f_ret = 0;
```
Comment (wangzhou) [inline test/hisi_zip_test/test_sva_perf.c]: seems one ret flag is enough

File: test/hisi_zip_test/test_sva_perf.c [hunk 10]
```diff
@@ -732,6 +733,54 @@ static void handle_sigbus(int sig)
...
+{
+	struct hizip_test_info info = {0};
+	struct test_options opts = {
+		.alg_type		= WD_ZLIB,
+		.sync_mode		= 0,
+		.thread_num		= 16,
+		.block_size		= 8192,
+		.total_len		= 8192 * 10,
+		.compact_run_num	= 1000,
+	};
```
Comment (wangzhou) [inline test/hisi_zip_test/test_sva_perf.c]: why not reuse the common input parameter? this breaks the whole test into two parts...

File: test/hisi_zip_test/test_sva_perf.c [hunk 11]
```diff
@@ -732,6 +733,54 @@ static void handle_sigbus(int sig)
...
+		return ret;
+	gettimeofday(&start_tvl, NULL);
+	ret = attach_threads(&opts, &info);
+	if (ret)
+		return ret;
+	gettimeofday(&end_tvl, NULL);
+	timersub(&end_tvl, &start_tvl, &start_tvl);
+	usec = (double)(start_tvl.tv_sec * 1000000 + start_tvl.tv_usec);
+	ilen = opts.total_len * opts.thread_num * opts.compact_run_num;
+	speed = ilen * 1000 * 1000 / 1024 / 1024 / usec;
```
Comment (wangzhou) [inline test/hisi_zip_test/test_sva_perf.c]: seems we can create a function to do this, some codes are in common ：）

File: test/hisi_zip_test/test_lib.h [hunk 12]
```diff
@@ -111,23 +130,23 @@ struct hizip_test_info {
...
 void *poll_thread_func(void *arg);
+void *sw_dfl_sw_ifl(void *arg);
 int create_send_threads(struct test_options *opts,
 			struct hizip_test_info *info,
 			void *(*send_thread_func)(void *arg)
 			);
+int create_send2_threads(struct test_options *opts,
+			 struct hizip_test_info *info,
+			 void *(*send_thread_func)(void *arg)
+			);
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.h:171]: should be  put in the tail of above line

File: test/hisi_zip_test/test_sva_perf.c [hunk 13]
```diff
@@ -732,6 +733,54 @@ static void handle_sigbus(int sig)
...
+	};
+	struct timeval start_tvl, end_tvl;
+	double ilen, usec, speed;
+	int ret;
+
+	info.opts = &opts;
+	ret = create_send2_threads(&opts, &info, sw_dfl_sw_ifl);
+	if (ret)
+		return ret;
+	gettimeofday(&start_tvl, NULL);
```
Comment (wangzhou) [inline test/hisi_zip_test/test_sva_perf.c]: start time should be got before "create_send2_threads"

File: test/hisi_zip_test/test_lib.c [hunk 14]
```diff
@@ -41,6 +42,239 @@ void *mmap_alloc(size_t len)
 	return p == MAP_FAILED ? NULL : p;
 }
 
+static void gen_random_data(void *buf, size_t len)
+{
+	int i;
+	uint32_t seed = 0;
+	unsigned short rand_state[3] = {(seed >> 16) & 0xffff, seed & 0xffff,
+					0x330e};
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c:66]: code style problem: better to put longer line above the shorter line here ：）

xxxxx
xxx
xx

File: test/hisi_zip_test/test_lib.c [hunk 15]
```diff
@@ -41,6 +42,239 @@ void *mmap_alloc(size_t len)
...
+static int cmp_md5(comp_md5_t *orig, comp_md5_t *final)
+{
+	int i;
+
+	if (!orig || !final)
+		return -EINVAL;
+	for (i = 0; i < MD5_DIGEST_LENGTH; i++) {
+		if (orig->md[i] != final->md[i]) {
+			printf("Original MD5: ");
+			dump_md5(orig);
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c:100]: seems this is for debug, is it useful to print it out if there is something wrong?

File: test/hisi_zip_test/test_lib.c [hunk 16]
```diff
@@ -41,6 +42,239 @@ void *mmap_alloc(size_t len)
...
+
+	deflateEnd(&strm);
+	return ret;
+}
+
+/*
+ * This function is used in BLOCK mode. Each compressing in BLOCK mode
+ * produces compression header.
+ */
+static int chunk_inflate(void *in, void *out, struct test_options *opts)
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c:228]: could we reuse what Jean had already done? I mean the software deflate/inflate function?

**Review by wangzhou — COMMENTED** [review]
test: zip: add cases for SW

What is the purpose to add this software case, if for the test of performance, it is better to do compress/decompress seperately

File: test/hisi_zip_test/test_sva_perf.c [hunk 17]
```diff
@@ -767,6 +767,54 @@ static int test_sw_dfl_sw_ifl(void)
...
+	struct hizip_test_info info = {0};
+	struct test_options opts = {
+		.alg_type		= WD_ZLIB,
+		.sync_mode		= 0,
+		.thread_num		= 16,
+		.q_num			= 16,
+		.block_size		= 8192,
+		.total_len		= 8192 * 10,
+		.compact_run_num	= 1000,
+	};
```
Comment (wangzhou) [inline test/hisi_zip_test/test_sva_perf.c]: same problem here as in sofware deflate/inflate

File: test/hisi_zip_test/test_sva_perf.c [hunk 18]
```diff
@@ -767,6 +767,54 @@ static int test_sw_dfl_sw_ifl(void)
...
+	info.list = get_dev_list(&opts, 1);
+	if (!info.list)
+		return -EINVAL;
+	ret = init_ctx_config(&opts, &info, &sched);
+	if (ret)
+		goto out;
+	ret = create_send2_threads(&opts, &info, sw_dfl_hw_ifl);
+	if (ret)
+		goto out_thd;
+	gettimeofday(&start_tvl, NULL);
```
Comment (wangzhou) [inline test/hisi_zip_test/test_sva_perf.c]: same problem here.

File: test/hisi_zip_test/test_lib.c [hunk 19]
```diff
@@ -275,6 +275,104 @@ void *sw_dfl_sw_ifl(void *arg)
...
+	struct wd_comp_sess_setup setup = {0};
+	handle_t h_ifl;
+	void *tbuf;
+	size_t tbuf_sz;
+	comp_md5_t final_md5;
+	int i, ret;
+
+        setup.alg_type = opts->alg_type;
+        setup.mode = opts->sync_mode ? CTX_MODE_ASYNC : CTX_MODE_SYNC;
+        setup.op_type = WD_DIR_DECOMPRESS;
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c]: coding style problem here?

File: test/hisi_zip_test/test_lib.c [hunk 20]
```diff
@@ -275,6 +275,104 @@ void *sw_dfl_sw_ifl(void *arg)
...
+		ret = calculate_md5(&final_md5, tdata->dst, tdata->dst_sz);
+		if (ret) {
+			printf("Fail to generate MD5 (%d)\n", ret);
+			goto out_run;
+		}
+		ret = cmp_md5(&tdata->md5, &final_md5);
+		if (ret) {
+			printf("MD5 is unmatched (%d)\n", ret);
+			goto out_run;
+		}
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c:694]: I just think that it makes no sense to get the performance data if you do md5 calculation here.

**Review by wangzhou — COMMENTED** [review]
same problems as in above patch

File: test/hisi_zip_test/test_lib.c [hunk 21]
```diff
@@ -617,6 +700,40 @@ void *hw_ifl_perf(void *arg)
...
+	struct timeval start_tvl, end_tvl;
+	int pending, local_sum = 0;
+
+	gettimeofday(&start_tvl, NULL);
+	while (sum_expect > total_recv) {
+		pending = __atomic_load_n(&sum_pend, __ATOMIC_ACQUIRE);
+		if (pending == 0)
+			continue;
+		received = 0;
+		ret = wd_comp_poll(pending, &received);
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c:689]: better to poll ctx directly

**Review by wangzhou — COMMENTED** [review]
test: zip: support async mode in self test

for this patch, you can refer to what we did in wd env, there should have a buffer to cache sent requests.

File: test/hisi_zip_test/test_sva_perf.c [hunk 22]
```diff
@@ -872,6 +886,7 @@ static int test_hw_dfl_hw_ifl(struct test_options *opts)
 	struct timeval start_tvl, end_tvl;
 	struct wd_sched *sched = NULL;
 	double ilen, usec, speed;
+	char zbuf[60];
```
Comment (wangzhou) [inline test/hisi_zip_test/test_sva_perf.c]: better to change to a macro to avoid magic number here :)

File: test/hisi_zip_test/test_sva_perf.c [hunk 23]
```diff
@@ -1020,57 +1056,59 @@ static int run_self_test(void)
...
+			opts.block_size = 8192; opts.total_len = 8192 * 10;
+			break;
+		case 3:
+			opts.sync_mode = 1;	opts.poll_num = 4;
+			opts.block_size = 8192;	opts.total_len = 8192 * 10;
+			break;
+		case 4:
+			opts.sync_mode = 1; 	opts.poll_num = 8;
+			opts.block_size = 8192; opts.total_len = 8192 * 10;
+			break;
```
Comment (wangzhou) [inline test/hisi_zip_test/test_sva_perf.c]: better to make them dynamically, and use a script to change different parameters :)

**Review by wangzhou — COMMENTED** [review]
test: zip: make poll thread numbers configurable

for this patch, better to have a comments somewhere to indicate the map relationship between polling threads and ctxs

File: test/hisi_zip_test/test_lib.c [hunk 24]
```diff
@@ -21,7 +21,7 @@ struct check_rand_ctx {
 
 static pthread_mutex_t mutex = PTHREAD_MUTEX_INITIALIZER;
 static int count = 0;
-static int sum_pend = 0, sum_expect = 0, sum_recv = 0;
+static int sum_pend = 0, sum_thread_end = 0;
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c]: better to store them somewhere to avoid global variables

File: test/hisi_zip_test/test_sva_perf.c [hunk 25]
```diff
@@ -1085,6 +1089,26 @@ static int run_self_test(void)
...
+			opts.block_size = 1024; opts.total_len = 8192 * 10;
+			break;
+		case 8:
+			opts.sync_mode = 1;	opts.poll_num = 4;
+			opts.block_size = 1024;	opts.total_len = 8192 * 10;
+			break;
+		case 9:
+			opts.sync_mode = 1; 	opts.poll_num = 8;
+			opts.block_size = 1024; opts.total_len = 8192 * 10;
+			break;
```
Comment (wangzhou) [inline test/hisi_zip_test/test_sva_perf.c]: better to make them dynamically, if you go through current block test, you will find it already support any size block size :)

File: test/hisi_zip_test/test_lib.c [hunk 26]
```diff
@@ -690,6 +816,88 @@ void *hw_ifl_perf(void *arg)
...
+	thread_data_t *tdata = (thread_data_t *)arg;
+	struct hizip_test_info *info = tdata->info;
+	struct test_options *opts = info->opts;
+	struct wd_comp_sess_setup setup = {0};
+	handle_t h_dfl;
+	int i, ret;
+
+        setup.alg_type = opts->alg_type;
+        setup.mode = opts->sync_mode ? CTX_MODE_ASYNC : CTX_MODE_SYNC;
+        setup.op_type = WD_DIR_COMPRESS;
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c]: code style problem here?

File: test/hisi_zip_test/test_lib.c [hunk 27]
```diff
@@ -690,6 +816,88 @@ void *hw_ifl_perf(void *arg)
...
+	thread_data_t *tdata = (thread_data_t *)arg;
+	struct hizip_test_info *info = tdata->info;
+	struct test_options *opts = info->opts;
+	struct wd_comp_sess_setup setup = {0};
+	handle_t h_ifl;
+	int i, ret;
+
+        setup.alg_type = opts->alg_type;
+        setup.mode = opts->sync_mode ? CTX_MODE_ASYNC : CTX_MODE_SYNC;
+        setup.op_type = WD_DIR_DECOMPRESS;
```
Comment (wangzhou) [inline test/hisi_zip_test/test_lib.c]: same as above

**Review by wangzhou — COMMENTED** [review]
test: zip: add batch test for async mode

for this patch, it seems that it does the same thing as what to do in -l, I mean compact mode.
Another idea is uadk support batch request, which need to add new uadk API.  The requirement of this case is simple: users offer uadk a batch of block requests to handle

**Review by wangzhou — COMMENTED** [review]
test: zip: move cases into testsuit 

for this patch and all above code.

In fact, what I expect is to add multiple thread test in current codes, instead of add a new module, which make our test code more split :(

File: test/hisi_zip_test/testsuit.c [hunk 28]
```diff
@@ -514,227 +514,117 @@ static int test_sw_dfl_sw_ifl(void)
...
+		ifl_flag = 1;
+	} else if (!strcmp(model, "hw_ifl_perf2")) {
+		func = hw_ifl_perf2;
+		info.in_size = opts->total_len * EXPANSION_RATIO;
+		info.out_size = opts->total_len;
+		zbuf_idx = sprintf(zbuf, "HW %s %s inflate",
+				   opts->sync_mode ? "ASYNC" : "SYNC",
+				   opts->is_stream ? "STREAM" : "BLOCK");
+		ifl_flag = 1;
+	} else {
```
Comment (wangzhou) [inline test/hisi_zip_test/testsuit.c:589]: above codes can be modify to a static table

**Comment by hzhuang1** [comment]
This PR depends on https://github.com/Linaro/uadk/pull/288.

**Pull request merged**